### ISSUE TITLE
[Brokernode] "busy buffer" error on brokernode

### ISSUE THREAD

**Issue by rfornea** [oysterprotocol/brokernode#280]
Getting a "busy buffer" error on the brokernode when trying to upload large files.  

https://oysterprotocol.slack.com/archives/C9VKK11PA/p1525968339000472

**Comment by AaronVasquez** [comment]
Could it be related to this? https://github.com/go-sql-driver/mysql/issues/314

Do you have a consistent way to reproduce this?

**Comment by automyr** [comment]
Could this actually be caused by the bulk inserts we introduced recently? 

It seems the issue resides in things still existing in the buffer while we're trying to use it for something else. Considering that row inserts in mariadb have a max size allowed and that this only happens on big files, I thought that maybe what's going on is that we're trying to insert too much data at once, it doesn't work but doesn't throw an immediate error, it just waits a little before trying to insert the rest of the data while other parts of the code see the successful insert and think the buffer must be empty.

Another related possibility could be that it only inserts what fits in a single operation, the rest of the buffer is not read and the insert operation just returns a success. Then the next time we try to do anything with the buffer, it's already filled with all the unread data and it fails.

I'm just throwing out wild guesses here as I've not tested anything, but perhaps it's related to something like this.

**Comment by AaronVasquez** [comment]
@pzhao5 any ideas? Bulk inserts might have introduced this? Are we batching the bulk inserts into smaller batches? I wonder if its using up too much memory if the bulk insert is too large...

**Comment by rfornea** [comment]
To be fair, we never tried such a big upload, even before the bulk inserts were in place.  This may have happened beforehand if we had used large enough files.

**Comment by pzhao5** [comment]
Let me break it into chunks

**Comment by pzhao5** [comment]
Submitted: https://github.com/oysterprotocol/brokernode/pull/282

Can you test it again?

**Comment by pzhao5** [comment]
this error was too general. Is there a way to verify that DB was updated?

**Comment by pzhao5** [comment]
Interesting. Yep, that OR statement could be large. I was not aware of that chunkID could be continuous

**Comment by pzhao5** [comment]
What if you update SQL_BATCH_SIZE to be 1 or 2. Will it still have any issue?

**Comment by pzhao5** [comment]
So, If I understand it correct:

Large file will affect current SELECT statement. The fix is to do range.
Either Large/Small fill will affect the UPDATE statement. Could play around with SQL_BATCH_SIZE

**Comment by rfornea** [comment]
>Large file will affect current SELECT statement. The fix is to do range.   
Yeah.  There's probably other ways we can fix it, this is just the first thing I tried.  

>Either Large/Small fill will affect the UPDATE statement. Could play around with SQL_BATCH_SIZE   
I'll try different batch sizes and see.  I think the problem is just that by the time the next update is occurring, the previous one hasn't finished, causing the deadlock.  

I'll make a PR with my current changes, and we can also experiment with different batch sizes.

**Comment by rfornea** [comment]
Moving this to "done" for now as I think it is fixed.
